sw/device/lib/testing/check.h

```diff
@@ -53,33 +53,45 @@
  * comparison `result_` value.
  *
- * @param[out] result_ Result of the comparison.
  * @param actual_ Buffer containing actual values.
  * @param expected_ Buffer containing expected values.
  * @param num_items_ Number of items to compare.
+ * @param ... Arguments to a LOG_* macro, which are evaluated if the check.
  */
-#define CHECK_BUFFER(result_, actual_, expected_, num_items_)                 \
-  do {                                                                        \
-    /* `sizeof(actual_[0])` is used to determine the size of each item in the \
-     * buffer. */                                                             \
-    if (memcmp(actual_, expected_, num_items_ * sizeof(actual_[0])) != 0) {   \
-      for (size_t i = 0; i < num_items_; ++i) {                               \
-        LOG_ERROR("[%d] actual = 0x%x; expected = 0x%x", i, actual_[i],       \
-                  expected_[i]);                                              \
-      }                                                                       \
-      result_ = false;                                                        \
-    } else {                                                                  \
-      result_ = true;                                                         \
-    }                                                                         \
+#define CHECK_BUFFER(actual_, expected_, num_items_, ...)                      \
+  do {                                                                         \
+    /* `sizeof(actual_[0])` is used to determine the size of each item in the  \
+     * buffer. */                                                              \
+    if (memcmp(actual_, expected_, num_items_ * sizeof(actual_[0])) != 0) {    \
+      if (GET_NUM_VARIABLE_ARGS(_, ##__VA_ARGS__) == 0) {                      \
+        LOG_ERROR("CHECK-BUFFER-fail: " #actual_ "does not match" #expected_); \
+      } else {                                                                 \
+        LOG_ERROR("CHECK-BUFFER-fail: " __VA_ARGS__);                          \
+      }                                                                        \
+      /* Note in DV, the simulation is terminated immediately upon the         \
+      execution of the first `LOG_ERROR(...)` statement, unless the            \
+      quit-on-error count is manually set to a higher value on the command     \
+      line. Therefore, the following `LOG_ERROR(...)` statements may not       \
+      execute in DV simulation.*/                                              \
+      for (size_t i = 0; i < num_items_; ++i) {                                \
+        LOG_ERROR("[%d] actual = 0x%x; expected = 0x%x", i, actual_[i],        \
+                  expected_[i]);                                               \
+      } /* Currently, this macro will call into                                \
```

Comment:
Perhaps the order could be swapped, and the loop in line 76 can be run as LOG_INFO, with the "failure" message being a LOG_ERROR? This is because it will be very informative to see the diffs.

Reply:
Fixed.

Note, DV sim will discard the LOG_INFO() prints, but they will be there in Verilator sims.

Reply:
DV sim does not discard any of the logs. It just uses a different mechanism to display the logs (completely bypassing UART).